src/workerd/io/worker.c++

```diff
@@ -978,6 +979,58 @@ kj::Maybe<kj::String> makeCompatJson(kj::ArrayPtr<kj::StringPtr> enableFlags) {
   return kj::String(json.releaseAsArray());
 }
 
+struct CrossThreadPromiseWaiter {
+  // When a promise is created in a different IoContext, we need to use a
+  // CrossThreadWaitList in order to wait on it. The Waiter instance will
+  // be held on the Promise itself, and will be fulfilled/rejected when the
+  // promise is resolved or rejected. This will signal all of the waiters
+  // from other IoContexts.
+  CrossThreadWaitList waitList;
```

Comment:
Instead of using `CrossThreadWaitList`, it seems like the code below would be simpler if you constructed a new `kj::newPromiseAndCrossThreadFulfiller()` pair each time the promise is cross-context awaited. Then you don't have to try to coordinate with any other waiter on the same process.

IIRC `CrossThreadWaitList` is essentially doing the same thing internally (creating a new pair for each waiter) so you aren't getting any performance benefit from trying to use it here.